emu_sv/custom_callback_implementations.py

```diff
@@ -147,3 +168,22 @@ def energy_second_moment_sv_impl(
     hstate = hamiltonian * state.vector
     en_2_mom: torch.Tensor = torch.vdot(hstate, hstate).real
     return en_2_mom.cpu()
+
+
+def energy_second_moment_den_mat_impl(
+    self: EnergyVariance,
+    *,
+    config: EmulationConfig,
+    state: DensityMatrix,
+    hamiltonian: RydbergLindbladian,
+) -> torch.Tensor:
+    """
+    Custom implementation of the second moment of energy tr(ρH²) for the
+    lindblad equation solver.
+    """
+    h_dense_matrix = hamiltonian.h_eff(state.matrix)  # Hρ
+    gpu = state.matrix.is_cuda
+
+    return hamiltonian.expect(
```

Comment:
doing `h_dense_matrix.overlap(h_dense_matrix)` will be faster. It saves an application of `hamiltonian`

Reply:
Mauro has alerted me to the fact that the default implementation of `EnergySecondMoment` and `EnergyVariance` in pulser are incorrect for density matrices. This makes the entire issue a bit moot. Let's leave the implementations as is. I will make a bugfix in pulser.